libraries/js/text/regex.effekt

```diff
@@ -1,33 +1,242 @@
+ *
+ * Each Group object contains:
+ * - matched: The matched substring for this group.
+ * - index: An Option[Range] with start and end indices if the 'd' flag is set, None otherwise.
+ *
+ * Note: This function is stateful when the global flag 'g' is set on the Regex object.
+ * Subsequent calls with the same Regex and string will return the next match in the string.
+ *
+ * Example usage:
+ * ```effekt
+ * val rx = regex("(\\w+)\\s(\\d+)", "g")
+ * val text = "Hello 123 World 456"
+ * while (true) {
+ *  exec(rx, text) match = {
+ *    Some(m) =>
+ *      println("matched: " ++ m.matched ++ "range: " ++ m.start.show() ++ ", " ++ m.end.show()))
+ *    None() => break
+ * }
+ * ```
+ */
+def exec(reg: Regex, str: String): Option[Match] = {
+  def toGroup(g: GroupMatch): Group = Group(
+    matched(g),
+    zip(undefinedToOption(start(g)), undefinedToOption(end(g))) { (s, e) => Range(s, e)}
+  )
+
+  val regexMatch = internal::unsafeExec(reg, str).undefinedToOption()
+  regexMatch.map { it =>
+    val fullMatch = fullMatch(it)
+    Match(
+      fullMatch.matched,
+      fullMatch.start,
+      fullMatch.end,
+      subGroups(it).toList().map{toGroup}
+    )
+  }
+}
+
+// internal representation
+extern type GroupMatch
+extern pure def matched(g: GroupMatch): String = js "${g}.matched"
+extern pure def start(g: GroupMatch): Int = js "${g}.start"
+extern pure def end(g: GroupMatch): Int = js "${g}.end"
 
-// internal representation { matched: String, index: Int }
 extern type RegexMatch
-extern pure def matched(r: RegexMatch): String = js "${r}.matched"
-extern pure def index(r: RegexMatch): Int = js "${r}.index"
+extern pure def fullMatch(r: RegexMatch): GroupMatch = js "${r}.fullMatch"
+extern pure def subGroups(r: RegexMatch): Array[GroupMatch] = js "${r}.groups"
 
 extern js """
 function regex$exec(reg, str) {
-  var res = reg.exec(str);
-  if (res === null) { return undefined }
-  else { return { matched: res[0], index: res.index } }
+  const match = reg.exec(str);
+  if (match === null)  return undefined
+
+  const fullMatch = { matched: match[0], start: match.index, end: reg.lastIndex };
+  var groups = [];
+  for (var i = 1; i < match.length; i++) {
+    if (match[i] !== undefined) {
+      if (match.indices !== undefined && match.indices[i] !== undefined) {
+        groups.push({
+          matched: match[i],
+          start: match.indices[i][0],
+          end: match.indices[i][1]
+        });
+      } else {
+        groups.push({
+          matched: match[i],
+        });
+      }
+    }else {
+      // Push an empty string if an intermediate group is undefined
+      groups.push({matched: ""});}
+  }
+
+  return { fullMatch: fullMatch, groups: groups };
+}
+
+"""
+
+// internal calls
+namespace internal {
+  extern io def unsafeExec(reg: Regex, str: String): RegexMatch = js "regex$exec(${reg}, ${str})"
+
+  extern pure def unsafeRegex(str: String, flags: String): Regex = js "new RegExp(${str}, ${flags})"
```

Comment:
This needs to be `io`, because otherwise the same regex will potentially be inlined over and over again and not have the "stateful" behavior.